### PR TITLE
Add get_global_ptr and improve linker functionality

### PR DESCRIPTION
**Pull request by benbarsdell**
Adds get_global_ptr as a more general version of get_constant_ptr, and improves linker support including better error logging and support for different file types. Relevant to https://github.com/NVIDIA/jitify/issues/46.

Also fixes a small bug in -remove-unused-globals.

Full details in commit log.

### CODE REVIEW

File: jitify_test.cpp [hunk 1]
```diff
@@ -688,6 +691,53 @@ TEST(JitifyTest, CuRandKernel) {
   // TODO: Expand this test to actually call curand kernels and check outputs.
 }
 
+static const char* const linktest_program1_source =
+    "linktest_program1\n"
+    "extern __constant__ int c = 5;\n"
+    "extern __device__ int d = 7;\n"
+    "extern __device__ int f(int i) { return i + 11; }\n"
+    "\n";
```
Comment (maddyscientist) [inline jitify_test.cpp]: Shouldn't these declarations not be `extern` here?  Or is this legal to have the definition marked `extern`?  I note that this works fine I remove the `extern` declarations as well.

**Comment by maddyscientist** [comment]
Great pull request adding some great functionality.  Just one question around https://github.com/NVIDIA/jitify/pull/47/commits/044705e4e066178e1e85959529d55549d713fe68 that I'd like answered before merging.

File: jitify_test.cpp (hunk 1 above)
Comment (benbarsdell) [reply jitify_test.cpp]: I admit I was unclear on the rules here. But this page says that `extern` is ignored on non-const definitions (and in fact required on const definitions):
https://docs.microsoft.com/en-us/cpp/cpp/extern-cpp?view=vs-2019#extern-linkage-for-non-const-globals
So I don't think it matters in this case (given that these are non-const). But I'll remove it if you think it's clearer that way.

Comment (benbarsdell) [reply jitify_test.cpp]: I've now removed them for clarity.

**Pull request merged**